File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

public class Solution {

    // 올바른 괄호 문자열인지 판별하는 메서드
    public static boolean solution(String s) {
        // 스택을 사용하여 괄호를 관리
        Deque<Character> stack = new ArrayDeque<>();

        // 문자열을 순회하며 검사
        for (char ch : s.toCharArray()) {
            // 여는 괄호 '('는 스택에 추가
            if (ch == '(') {
                stack.push(ch);
            } 
            // 닫는 괄호 ')'가 나오면 처리
            else {
                // 스택이 비어 있으면 올바르지 않은 괄호
                if (stack.isEmpty()) {
                    return false;
                }
                // 스택에서 '(' 제거
                stack.pop();
            }
        }

        // 스택이 비어 있으면 올바른 괄호, 그렇지 않으면 잘못된 괄호
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        // ✅ 예제 테스트 실행
        System.out.println(solution("()()"));     // true
        System.out.println(solution("(())()"));   // true
        System.out.println(solution(")()("));     // false
        System.out.println(solution("(()("));     // false
        System.out.println(solution("()"));       // true
        System.out.println(solution(")("));       // false
        System.out.println(solution(")"));        // false
        System.out.println(solution("("));        // false
        System.out.println(solution("))"));       // false
        System.out.println(solution("(("));       // false
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s) {
    // 괄호 짝을 판단할 스택 초기화
    let stack = [];

    // 문자열을 처음부터 끝까지 순회
    for (let char of s) {
        // 여는 괄호는 스택에 추가
        if (char === '(') {
            stack.push(char);
        }
        // 닫는 괄호가 나왔을 경우
        else if (char === ')') {
            // 스택이 비어 있다면 짝이 맞지 않음
            if (stack.length === 0) {
                return false;
            }
            // 가장 최근의 여는 괄호 제거
            stack.pop();
        }
    }

    // 스택이 비어 있으면 모든 괄호가 올바르게 닫힘 → true 반환
    // 스택이 남아 있으면 일부 괄호가 닫히지 않음 → false 반환
    return stack.length === 0;
}

// ✅ 예제 테스트 실행
console.log(solution("()()"));     // true
console.log(solution("(())()"));   // true
console.log(solution(")()("));     // false
console.log(solution("(()("));     // false
console.log(solution("()"));       // true
console.log(solution(")("));       // false
console.log(solution(")"));        // false
console.log(solution("("));        // false
console.log(solution("))"));       // false
console.log(solution("(("));       // false
